File: 2021/moons-and-umbrellas/typescript/main.ts
import readline from 'readline';

let cases = 0;

const reader = readline.createInterface({
    input: process.stdin,
});

reader.on('line', (line: string) => {
    const parts = line.split(' ')
    const mural = parts[2];

    let cost = 0;
    // First line is the number of cases, the array will be empty, therefore
    // `mural` is undefined
    if (mural) {
        const x = Number(parts[0]);
        const y = Number(parts[1]);
        let prev = '';
        [...mural].forEach(letter => {
            switch (letter) {
                case 'C':
                    if (prev === 'J') {
                        cost += y;
                    }
                    prev = letter
                    break;
                case 'J':
                    if (prev === 'C') {
                        cost += x;
                    }
                    prev = letter
                    break;
                default: {
                    // Ignore '?'
                }
            }
        })
        cases++
        console.log(`Case #${cases}: ${cost}`)
    }
})

File: 2021/moons-and-umbrellas/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	scanner := bufio.NewScanner(os.Stdin)

	// Skip number of tests
	scanner.Scan()

	var cases int
	
	for scanner.Scan() {
		parts := strings.Split(scanner.Text(), " ")
		
		x, _ := strconv.Atoi(parts[0])
		y, _ := strconv.Atoi(parts[1])
		mural := parts[2]
		
		var cost int
		var prev rune
		
		for _, letter := range mural {
			switch letter {
			case 'C':
				{

					if prev == 'J' {
						cost += y
					}
					prev = letter
				}
			case 'J':
				{

					if prev == 'C' {
						cost += x
					}
					prev = letter
				}
			default:
				{
					// Ignore '?'
				}
			}
		}
		cases++
		fmt.Printf("Case #%d: %d\n", cases, cost)
	}
}
